`pkgs.dwarf-fortress-packages.dwarf-fortress-full` does not build on unstable

### Describe the bug
See log below.

### Steps To Reproduce
Steps to reproduce the behavior:
1. `sudo nixos-rebuild switch --upgrade-all`

### Expected behavior
Should build.

### Additional context

Log:

```console
make[2]: *** [plugins/CMakeFiles/autogems.dir/build.make:76: plugins/CMakeFiles/autogems.dir/autogems.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1686: plugins/CMakeFiles/autogems.dir/all] Error 2
In file included from /build/source/plugins/buildingplan-planner.cpp:16:
/build/source/plugins/uicommon.h: In function 'std::string int_to_string(int)':
/build/source/plugins/uicommon.h:163:59: error: taking address of rvalue [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-fpermissive-fpermissive8;;]
  163 |     return static_cast<ostringstream*>( &(ostringstream() << n) )->str();
      |                                          ~~~~~~~~~~~~~~~~~^~~~~
[…]
error: builder for '/nix/store/dick2v6psqpryw4a8qz625q6jzjad4wm-dfhack-base-0.47.05-r1.drv' failed with exit code 2
error: 1 dependencies of derivation '/nix/store/r3n2a4hidrxh6qn8y4lx15fackxiyqbx-dfhack-0.47.05-r1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/f3cx1qn99azh7lkk9j35md44n5hcn6dv-dwarf-fortress-env-0.47.05.drv' failed to build
error: 1 dependencies of derivation '/nix/store/c5a58xfzhkcyd8y0xa7yyb274gnz7088-dwarf-fortress-init.drv' failed to build
error: 1 dependencies of derivation '/nix/store/w36lv1yygs67bc3ylkyyii91hvsdd180-dwarf-fortress-0.47.05.drv' failed to build
error: 1 dependencies of derivation '/nix/store/z0fdqhsl9yv4wr8y0xqlkjgrlbl1w9wl-dwarf-fortress-full.drv' failed to build
error: 1 dependencies of derivation '/nix/store/gpmp9wqr008lgvg8i5w1ld5f5c54s7m8-user-environment.drv' failed to build
error: 1 dependencies of derivation '/nix/store/851dcij9q1mdshq8s1rdv2f7vacwg2jv-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/7af7qy0igyk0fybvmgrbhbfc02a22ws1-nixos-system-big-22.05pre372658.6a323903ad0.drv' failed to build
```

### Notify maintainers
@Baughn @numinit

### Metadata
```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.34, NixOS, 22.05 (Quokka), 22.05pre369988.1ffba9f2f68`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0`
 - channels(root): `"nixos-22.05pre372658.6a323903ad0"`
 - channels(victor): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
 ```

--- a/pkgs/games/dwarf-fortress/dfhack/default.nix
+++ b/pkgs/games/dwarf-fortress/dfhack/default.nix
@@ -131,6 +131,9 @@ let
 
     patches = [ ./fix-stonesense.patch ];
 
+    # gcc 11 fix
+    NIX_CFLAGS_COMPILE = "-fpermissive";
+
     # As of
     # https://github.com/DFHack/dfhack/commit/56e43a0dde023c5a4595a22b29d800153b31e3c4,
     # dfhack gets its goodies from the directory above the Dwarf_Fortress

